File: Examples/CSharp/x-IMU3-Examples/Examples/Connection.cs
namespace Ximu3Examples
{
    class Connection
    {
        protected static void Run(Ximu3.ConnectionConfig config)
        {
            // Create connection
            Ximu3.Connection connection = new(config);

            connection.AddReceiveErrorCallback(ReceiveErrorCallback);
            connection.AddStatisticsCallback(StatisticsCallback);

            if (Helpers.YesOrNo("Print data messages?"))
            {
                connection.AddInertialCallback(InertialCallback);
                connection.AddMagnetometerCallback(MagnetometerCallback);
                connection.AddQuaternionCallback(QuaternionCallback);
                connection.AddRotationMatrixCallback(RotationMatrixCallback);
                connection.AddEulerAnglesCallback(EulerAnglesCallback);
                connection.AddLinearAccelerationCallback(LinearAccelerationCallback);
                connection.AddEarthAccelerationCallback(EarthAccelerationCallback);
                connection.AddAhrsStatusCallback(AhrsStatusCallback);
                connection.AddHighGAccelerometerCallback(HighGAccelerometerCallback);
                connection.AddTemperatureCallback(TemperatureCallback);
                connection.AddBatteryCallback(BatteryCallback);
                connection.AddRssiCallback(RssiCallback);
                connection.AddSerialAccessoryCallback(SerialAccessoryCallback);
                connection.AddNotificationCallback(NotificationCallback);
                connection.AddErrorCallback(ErrorCallback);
            }
            connection.AddEndOfFileCallback(EndOfFileCallback);

            // Open connection
            Ximu3.CApi.XIMU3_Result result = connection.Open();

            if (result != Ximu3.CApi.XIMU3_Result.XIMU3_ResultOk)
            {
                Console.WriteLine("Unable to open " + connection.GetConfig() + ". " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)) + ".");
                return;
            }

            // Send command to strobe LED
            connection.SendCommand("{\"strobe\":null}");

            // Close connection
            System.Threading.Thread.Sleep(60000);

            connection.Close();
        }

        private static string TimestampFormat(UInt64 timestamp)
        {
            return $"{timestamp,8} us";
        }

        private static string IntFormat(uint value)
        {
            return $" {value,8}";
        }

        private static string IntFormat(ulong value)
        {
            return $" {value,8}";
        }

        private static string FloatFormat(float value)
        {
            return $" {value.ToString("###0.000", System.Globalization.CultureInfo.InvariantCulture),8}";
        }

        private static string StringFormat(string value)
        {
            return $@" ""{value}""";
        }

        private static void ReceiveErrorCallback(Ximu3.CApi.XIMU3_ReceiveError error)
        {
            Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_receive_error_to_string(error)));
        }

        private static void StatisticsCallback(Ximu3.CApi.XIMU3_Statistics statistics)
        {
            Console.WriteLine(TimestampFormat(statistics.timestamp) +
                              IntFormat(statistics.data_total) + " bytes" +
                              IntFormat(statistics.data_rate) + " bytes/s" +
                              IntFormat(statistics.message_total) + " messages" +
                              IntFormat(statistics.message_rate) + " messages/s" +
                              IntFormat(statistics.error_total) + " errors" +
                              IntFormat(statistics.error_rate) + " errors/s");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_statistics_to_string(statistics))); // alternative to above
        }

        private static void InertialCallback(Ximu3.CApi.XIMU3_InertialMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.gyroscope_x) + " deg/s" +
                              FloatFormat(message.gyroscope_y) + " deg/s" +
                              FloatFormat(message.gyroscope_z) + " deg/s" +
                              FloatFormat(message.accelerometer_x) + " g" +
                              FloatFormat(message.accelerometer_y) + " g" +
                              FloatFormat(message.accelerometer_z) + " g");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_inertial_message_to_string(message))); // alternative to above
        }

        private static void MagnetometerCallback(Ximu3.CApi.XIMU3_MagnetometerMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.x) + " a.u." +
                              FloatFormat(message.y) + " a.u." +
                              FloatFormat(message.z) + " a.u.");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_magnetometer_message_to_string(message))); // alternative to above
        }

        private static void QuaternionCallback(Ximu3.CApi.XIMU3_QuaternionMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.w) +
                              FloatFormat(message.x) +
                              FloatFormat(message.y) +
                              FloatFormat(message.z));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_quaternion_message_to_string(message))); // alternative to above
            Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_euler_angles_message_to_string(Ximu3.CApi.XIMU3_quaternion_message_to_euler_angles_message(message))));
        }

        private static void RotationMatrixCallback(Ximu3.CApi.XIMU3_RotationMatrixMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.xx) +
                              FloatFormat(message.xy) +
                              FloatFormat(message.xz) +
                              FloatFormat(message.yx) +
                              FloatFormat(message.yy) +
                              FloatFormat(message.yz) +
                              FloatFormat(message.zx) +
                              FloatFormat(message.zy) +
                              FloatFormat(message.zz));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_rotation_matrix_message_to_string(message))); // alternative to above
            Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_euler_angles_message_to_string(Ximu3.CApi.XIMU3_rotation_matrix_message_to_euler_angles_message(message))));
        }

        private static void EulerAnglesCallback(Ximu3.CApi.XIMU3_EulerAnglesMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.roll) + " deg" +
                              FloatFormat(message.pitch) + " deg" +
                              FloatFormat(message.yaw) + " deg");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_euler_angles_message_to_string(message))); // alternative to above
        }

        private static void LinearAccelerationCallback(Ximu3.CApi.XIMU3_LinearAccelerationMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.quaternion_w) +
                              FloatFormat(message.quaternion_x) +
                              FloatFormat(message.quaternion_y) +
                              FloatFormat(message.quaternion_z) +
                              FloatFormat(message.acceleration_x) + " g" +
                              FloatFormat(message.acceleration_y) + " g" +
                              FloatFormat(message.acceleration_z) + " g");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_linear_acceleration_message_to_string(message))); // alternative to above
            Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_euler_angles_message_to_string(Ximu3.CApi.XIMU3_linear_acceleration_message_to_euler_angles_message(message))));
        }

        private static void EarthAccelerationCallback(Ximu3.CApi.XIMU3_EarthAccelerationMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.quaternion_w) +
                              FloatFormat(message.quaternion_x) +
                              FloatFormat(message.quaternion_y) +
                              FloatFormat(message.quaternion_z) +
                              FloatFormat(message.acceleration_x) + " g" +
                              FloatFormat(message.acceleration_y) + " g" +
                              FloatFormat(message.acceleration_z) + " g");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_earth_acceleration_message_to_string(message))); // alternative to above
            Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_euler_angles_message_to_string(Ximu3.CApi.XIMU3_earth_acceleration_message_to_euler_angles_message(message))));
        }

        private static void AhrsStatusCallback(Ximu3.CApi.XIMU3_AhrsStatusMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.initialising) +
                              FloatFormat(message.angular_rate_recovery) +
                              FloatFormat(message.acceleration_recovery) +
                              FloatFormat(message.magnetic_recovery));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_ahrs_status_message_to_string(message))); // alternative to above
        }

        private static void HighGAccelerometerCallback(Ximu3.CApi.XIMU3_HighGAccelerometerMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.x) + " g" +
                              FloatFormat(message.y) + " g" +
                              FloatFormat(message.z) + " g");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_high_g_accelerometer_message_to_string(message))); // alternative to above
        }

        private static void TemperatureCallback(Ximu3.CApi.XIMU3_TemperatureMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.temperature) + " degC");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_temperature_message_to_string(message))); // alternative to above
        }

        private static void BatteryCallback(Ximu3.CApi.XIMU3_BatteryMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.percentage) + " %" +
                              FloatFormat(message.voltage) + " V" +
                              FloatFormat(message.charging_status) + $" ({Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_charging_status_to_string(Ximu3.CApi.XIMU3_charging_status_from_float(message.charging_status)))})");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_battery_message_to_string(message)) + $" ({Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_charging_status_to_string(Ximu3.CApi.XIMU3_charging_status_from_float(message.charging_status)))})"); // alternative to above
        }

        private static void RssiCallback(Ximu3.CApi.XIMU3_RssiMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) +
                              FloatFormat(message.percentage) + " %" +
                              FloatFormat(message.power) + " dBm");
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_rssi_message_to_string(message))); // alternative to above
        }

        private static void SerialAccessoryCallback(Ximu3.CApi.XIMU3_SerialAccessoryMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) + StringFormat(Ximu3.Helpers.ToString(message.char_array)));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_serial_accessory_message_to_string(message))); // alternative to above
        }

        private static void NotificationCallback(Ximu3.CApi.XIMU3_NotificationMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) + StringFormat(Ximu3.Helpers.ToString(message.char_array)));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_notification_message_to_string(message))); // alternative to above
        }

        private static void ErrorCallback(Ximu3.CApi.XIMU3_ErrorMessage message)
        {
            Console.WriteLine(TimestampFormat(message.timestamp) + StringFormat(Ximu3.Helpers.ToString(message.char_array)));
            // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_error_message_to_string(message))); // alternative to above
        }

        private static void EndOfFileCallback()
        {
            Console.WriteLine("End of file");
        }
    }
}

File: Examples/Python/connection.py
import time

import helpers
import ximu3


def timestamp_format(timestamp: int) -> str:
    return f"{timestamp:8d} us"


def int_format(value: int) -> str:
    return f" {value:8d}"


def float_format(value: float) -> str:
    return f" {value:8.3f}"


def string_format(string: str) -> str:
    return f' "{string}"'


def receive_error_callback(error: int) -> None:
    print(ximu3.receive_error_to_string(error))


def statistics_callback(statistics: ximu3.Statistics) -> None:
    print(
        "".join(
            [
                timestamp_format(statistics.timestamp),
                int_format(statistics.data_total) + " bytes",
                int_format(statistics.data_rate) + " bytes/s",
                int_format(statistics.message_total) + " messages",
                int_format(statistics.message_rate) + " messages/s",
                int_format(statistics.error_total) + " errors",
                int_format(statistics.error_rate) + " errors/s",
            ]
        )
    )
    # print(statistics)  # alternative to above


def inertial_callback(message: ximu3.InertialMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.gyroscope_x) + " deg/s",
                float_format(message.gyroscope_y) + " deg/s",
                float_format(message.gyroscope_z) + " deg/s",
                float_format(message.accelerometer_x) + " g",
                float_format(message.accelerometer_y) + " g",
                float_format(message.accelerometer_z) + " g",
            ]
        )
    )
    # print(message)  # alternative to above


def magnetometer_callback(message: ximu3.MagnetometerMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.x) + " a.u.",
                float_format(message.y) + " a.u.",
                float_format(message.z) + " a.u.",
            ]
        )
    )
    # print(message)  # alternative to above


def high_g_accelerometer_callback(message: ximu3.HighGAccelerometerMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.x) + " g",
                float_format(message.y) + " g",
                float_format(message.z) + " g",
            ]
        )
    )
    # print(message)  # alternative to above


def quaternion_callback(message: ximu3.QuaternionMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.w),
                float_format(message.x),
                float_format(message.y),
                float_format(message.z),
            ]
        )
    )
    # print(message)  # alternative to above
    print(message.to_euler_angles_message())


def rotation_matrix_callback(message: ximu3.RotationMatrixMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.xx),
                float_format(message.xy),
                float_format(message.xz),
                float_format(message.yx),
                float_format(message.yy),
                float_format(message.yz),
                float_format(message.zx),
                float_format(message.zy),
                float_format(message.zz),
            ]
        )
    )
    # print(message)  # alternative to above
    print(message.to_euler_angles_message())


def euler_angles_callback(message: ximu3.EulerAnglesMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.roll) + " deg",
                float_format(message.pitch) + " deg",
                float_format(message.yaw) + " deg",
            ]
        )
    )
    # print(message)  # alternative to above
    print(message.to_quaternion_message())


def linear_acceleration_callback(message: ximu3.LinearAccelerationMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.quaternion_w),
                float_format(message.quaternion_x),
                float_format(message.quaternion_y),
                float_format(message.quaternion_z),
                float_format(message.acceleration_x) + " g",
                float_format(message.acceleration_y) + " g",
                float_format(message.acceleration_z) + " g",
            ]
        )
    )
    # print(message)  # alternative to above
    print(message.to_euler_angles_message())


def earth_acceleration_callback(message: ximu3.EarthAccelerationMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.quaternion_w),
                float_format(message.quaternion_x),
                float_format(message.quaternion_y),
                float_format(message.quaternion_z),
                float_format(message.acceleration_x) + " g",
                float_format(message.acceleration_y) + " g",
                float_format(message.acceleration_z) + " g",
            ]
        )
    )
    # print(message)  # alternative to above
    print(message.to_euler_angles_message())


def ahrs_status_callback(message: ximu3.AhrsStatusMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.initialising),
                float_format(message.angular_rate_recovery),
                float_format(message.acceleration_recovery),
                float_format(message.magnetic_recovery),
            ]
        )
    )
    # print(message)  # alternative to above


def serial_accessory_callback(message: ximu3.SerialAccessoryMessage) -> None:
    print(timestamp_format(message.timestamp) + string_format(message.string))
    # print(message)  # alternative to above


def sync_callback(message: ximu3.SyncMessage) -> None:
    print(timestamp_format(message.timestamp) + float_format(message.edge))
    # print(message)  # alternative to above


def ltc_callback(message: ximu3.LtcMessage) -> None:
    print(timestamp_format(message.timestamp) + string_format(message.string))
    # print(message)  # alternative to above


def temperature_callback(message: ximu3.TemperatureMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.temperature) + " degC",
            ]
        )
    )
    # print(message)  # alternative to above


def battery_callback(message: ximu3.BatteryMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.percentage) + " %",
                float_format(message.voltage) + " V",
                float_format(message.charging_status) + f" ({ximu3.charging_status_to_string(ximu3.charging_status_from_float(message.charging_status))})",
            ]
        )
    )
    # print(message)  # alternative to above


def rssi_callback(message: ximu3.RssiMessage) -> None:
    print(
        "".join(
            [
                timestamp_format(message.timestamp),
                float_format(message.percentage) + " %",
                float_format(message.power) + " dBm",
            ]
        )
    )
    # print(message)  # alternative to above


def button_callback(message: ximu3.ButtonMessage) -> None:
    print(timestamp_format(message.timestamp) + float_format(message.state))
    # print(message)  # alternative to above


def notification_callback(message: ximu3.NotificationMessage) -> None:
    print(timestamp_format(message.timestamp) + string_format(message.string))
    # print(message)  # alternative to above


def error_callback(message: ximu3.ErrorMessage) -> None:
    print(timestamp_format(message.timestamp) + string_format(message.string))
    # print(message)  # alternative to above


def end_of_file_callback() -> None:
    print("End of file")


def run(config: ximu3.ConnectionConfig) -> None:
    connection = ximu3.Connection(config)

    connection.open()

    connection.send_command('{"strobe":null}')  # send command to strobe LED

    if helpers.yes_or_no("Use callbacks (else poll)?"):
        connection.add_receive_error_callback(receive_error_callback)
        connection.add_statistics_callback(statistics_callback)
        connection.add_inertial_callback(inertial_callback)
        connection.add_magnetometer_callback(magnetometer_callback)
        connection.add_high_g_accelerometer_callback(high_g_accelerometer_callback)
        connection.add_quaternion_callback(quaternion_callback)
        connection.add_rotation_matrix_callback(rotation_matrix_callback)
        connection.add_euler_angles_callback(euler_angles_callback)
        connection.add_linear_acceleration_callback(linear_acceleration_callback)
        connection.add_earth_acceleration_callback(earth_acceleration_callback)
        connection.add_ahrs_status_callback(ahrs_status_callback)
        connection.add_serial_accessory_callback(serial_accessory_callback)
        connection.add_sync_callback(sync_callback)
        connection.add_ltc_callback(ltc_callback)
        connection.add_temperature_callback(temperature_callback)
        connection.add_battery_callback(battery_callback)
        connection.add_rssi_callback(rssi_callback)
        connection.add_button_callback(button_callback)
        connection.add_notification_callback(notification_callback)
        connection.add_error_callback(error_callback)
        connection.add_end_of_file_callback(end_of_file_callback)

        time.sleep(60)
    else:
        for _ in range(60000):
            print(connection.get_statistics())
            print(connection.get_inertial_message())
            print(connection.get_magnetometer_message())
            print(connection.get_high_g_accelerometer_message())
            print(connection.get_quaternion_message())
            print(connection.get_rotation_matrix_message())
            print(connection.get_euler_angles_message())
            print(connection.get_linear_acceleration_message())
            print(connection.get_earth_acceleration_message())
            print(connection.get_ahrs_status_message())
            print(connection.get_serial_accessory_message())
            print(connection.get_sync_message())
            print(connection.get_ltc_message())
            print(connection.get_temperature_message())
            print(connection.get_battery_message())
            print(connection.get_rssi_message())
            print(connection.get_button_message())
            print(connection.get_notification_message())
            print(connection.get_error_message())

            time.sleep(0.001)

    connection.close()
